### PR TITLE
feat: add ext/node for require support

### PR DESCRIPTION
**Pull request by bartlomieju**
This commit adds "ext/node" extension that implementes CommonJS module system.

In the future this extension might be extended to actually contain implementation of
Node compatibility layer in favor of "deno_std/node".

Currently this functionality is not publicly exposed, it is available via "Deno[Deno.internal].require"
namespace and is meant to be used by other functionality to be landed soon.

This is a minimal first pass, things that still don't work:
- support for dynamic imports in CJS
- conditional exports

### CODE REVIEW

File: cli/compat/mod.rs [hunk 1]
```diff
@@ -147,21 +170,6 @@ fn escape_for_single_quote_string(text: &str) -> String {
   text.replace('\\', r"\\").replace('\'', r"\'")
 }
 
-pub fn setup_builtin_modules(
```
Comment (bartlomieju) [inline cli/compat/mod.rs:150]: Revert before landing

File: cli/main.rs [hunk 2]
```diff
@@ -869,8 +869,7 @@ async fn repl_command(
   if ps.options.compat() {
     worker.execute_side_module(&compat::GLOBAL_URL).await?;
     compat::add_global_require(&mut worker.js_runtime, main_module.as_str())?;
-    worker.run_event_loop(false).await?;
-    compat::setup_builtin_modules(&mut worker.js_runtime)?;
+    compat::load_builtin_node_modules(&mut worker.js_runtime).await?;
```
Comment (bartlomieju) [inline cli/main.rs:872]: Revert before landing

File: cli/tools/bench.rs [hunk 3]
```diff
@@ -381,6 +381,7 @@ async fn bench_specifier(
     if use_esm_loader {
       worker.execute_side_module(&specifier).await?;
     } else {
+      compat::load_builtin_node_modules(&mut worker.js_runtime).await?;
```
Comment (bartlomieju) [inline cli/tools/bench.rs:384]: Revert before landing

File: cli/tools/test.rs [hunk 4]
```diff
@@ -775,6 +775,7 @@ async fn test_specifier(
       if use_esm_loader {
         worker.execute_side_module(&specifier).await?;
       } else {
+        compat::load_builtin_node_modules(&mut worker.js_runtime).await?;
```
Comment (bartlomieju) [inline cli/tools/test.rs:778]: Revert before landing

File: ext/node/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,17 @@
+# Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
+
+[package]
+name = "deno_node"
+version = "0.0.1"
```
Comment (bartlomieju) [inline ext/node/Cargo.toml:5]: Bump to `0.1.0` before landing an publish to crates.io

File: ext/node/lib.rs [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+  p.parent().unwrap().to_string_lossy().to_string()
+}
+
+#[op]
+fn op_require_path_basename(request: String) -> String {
+  let p = PathBuf::from(request);
+  p.file_name().unwrap().to_string_lossy().to_string()
+}
+
+#[op]
```
Comment (bartlomieju) [inline ext/node/lib.rs:266]: Remove

File: require_test/example.js [hunk 7]
```diff
@@ -0,0 +1,11 @@
+// Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
```
Comment (bartlomieju) [inline require_test/example.js:1]: Move that directory to `cli/tests/testdata/require` or `cli/tests/testdata/commonjs`

**Review by dsherret — APPROVED** [review]
LGTM!

File: ext/node/lib.rs [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+      if let Ok(cwd) = std::env::current_dir() {
+        return Some(cwd.to_string_lossy().to_string());
+      }
+    }
+  }
+  None
+}
+
+#[op]
+fn op_require_read_file(filename: String) -> Result<String, AnyError> {
```
Comment (dsherret) [inline ext/node/lib.rs:314]: Missing permission checks. Will open a PR to fix.

**Pull request merged**